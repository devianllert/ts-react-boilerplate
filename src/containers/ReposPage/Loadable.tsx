import loadable, { ImportComponent } from '../../utils/loadable';

const LoadableCounter = loadable((): ImportComponent => import(/* webpackChunkName: "reposPage" */ './ReposPage'));

export default LoadableCounter;
