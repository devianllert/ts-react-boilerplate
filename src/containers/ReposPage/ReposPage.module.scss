@import '../../design/design-system';

.content {
  display: flex;

  flex-direction: column;
  
  align-items: center;
  justify-content: center;

  flex: 1;
}

.counter {
  display: flex;

  align-items: center;
}

.count {
  font-size: $font-size-h3;
  
  line-height: $line-height-h3;
}

.info {
  margin-top: $space-16;

  font-size: $font-size;

  line-height: $line-height;

  color: var(--brand-text-secondary-color)
}