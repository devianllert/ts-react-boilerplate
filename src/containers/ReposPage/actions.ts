import {
  ReposRequest,
  ReposSuccess,
  ReposFailure,
  ReposFulfill,
  ReposTrigger,
} from './types';
import {
  REPOS_REQUEST,
  REPOS_FAILURE,
  REPOS_SUCCESS,
  REPOS_FULFILL,
  REPOS_TRIGGER,
} from './constants';
import { Repo } from '../../services/repos.service';

const reposTrigger = (): ReposTrigger => ({
  type: REPOS_TRIGGER,
});

const reposRequest = (): ReposRequest => ({
  type: REPOS_REQUEST,
});

const reposSuccess = (data: Repo[]): ReposSuccess => ({
  type: REPOS_SUCCESS,
  payload: data,
});

const reposFailure = (error: Error): ReposFailure => ({
  type: REPOS_FAILURE,
  payload: error,
});

const reposFulfill = (): ReposFulfill => ({
  type: REPOS_FULFILL,
});

export {
  reposTrigger,
  reposRequest,
  reposSuccess,
  reposFailure,
  reposFulfill,
};
