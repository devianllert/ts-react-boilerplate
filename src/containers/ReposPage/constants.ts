export const REPOS_TRIGGER = 'boilerplate/Repos/REPOS_TRIGGER';
export const REPOS_REQUEST = 'boilerplate/Repos/REPOS_REQUEST';
export const REPOS_SUCCESS = 'boilerplate/Repos/REPOS_SUCCESS';
export const REPOS_FAILURE = 'boilerplate/Repos/REPOS_FAILURE';
export const REPOS_FULFILL = 'boilerplate/Repos/REPOS_FULFILL';
