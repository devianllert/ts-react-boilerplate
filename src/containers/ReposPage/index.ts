import Repos from './ReposPage';

export default Repos;
