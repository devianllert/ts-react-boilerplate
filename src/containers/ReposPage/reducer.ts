import { Reducer } from 'redux';

import { ReposState, ReposActions } from './types';
import {
  REPOS_REQUEST,
  REPOS_FAILURE,
  REPOS_SUCCESS,
  REPOS_FULFILL,
} from './constants';

export const initialState: ReposState = {
  data: [],
  loading: false,
};

const reposReducer: Reducer<ReposState, ReposActions> = (state = initialState, action): ReposState => {
  switch (action.type) {
    case REPOS_REQUEST:
      return {
        ...state,
        loading: true,
      };

    case REPOS_FAILURE:
      return {
        ...state,
        error: action.payload,
      };

    case REPOS_SUCCESS:
      return {
        ...state,
        data: [...action.payload],
      };

    case REPOS_FULFILL:
      return {
        ...state,
        loading: false,
      };

    default:
      return state;
  }
};

export default reposReducer;
