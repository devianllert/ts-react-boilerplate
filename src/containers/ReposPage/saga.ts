import { takeLatest, put, call } from 'redux-saga/effects';

import {
  reposRequest,
  reposSuccess,
  reposFailure,
  reposFulfill,
} from './actions';
import { REPOS_TRIGGER } from './constants';
import { ReposTrigger } from './types';

import { fetchRepos } from '../../services/repos.service';

export function* getRepos(action: ReposTrigger) {
  try {
    yield put(reposRequest());

    const data = yield call(fetchRepos, action.payload);

    yield put(reposSuccess(data));
  } catch (err) {
    yield put(reposFailure(err));
  } finally {
    yield put(reposFulfill());
  }
}

export default function* reposData() {
  yield takeLatest(REPOS_TRIGGER, getRepos);
}
