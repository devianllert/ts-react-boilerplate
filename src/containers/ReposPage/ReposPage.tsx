import React, { ReactElement, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { Tooltip } from 'react-essential-tools';

import { ReposTrigger } from './types';

import { reposTrigger } from './actions';
import { makeSelectReposData } from './selectors';

import useInjectReducer from '../../hooks/useInjectReducer';
import useInjectSaga from '../../hooks/useInjectSaga';

import reducer from './reducer';
import saga from './saga';

import Button from '../../components/Button';

import styles from './ReposPage.module.scss';

const ReposPage = (): ReactElement => {
  useInjectReducer({ key: 'repos', reducer });
  useInjectSaga({ key: 'repos', saga });

  const { t } = useTranslation();
  const repos = useSelector(makeSelectReposData());
  const dispatch = useDispatch();

  const [name, setName] = useState('');

  const requestRepos = (): ReposTrigger => dispatch(reposTrigger(name));


  return (
    <>
      <Helmet>
        <title>{t('counter.headTitle')}</title>
      </Helmet>

      <div className={styles.content}>
        <input
          type="text"
          placeholder="Enter repository name"
          name="repository"
          value={name}
          onChange={(event): void => setName(event.target.value)}
        />

        <Tooltip title="Fetch repositories">
          <Button onClick={requestRepos}>Fetch</Button>
        </Tooltip>

        {(repos.length > 0) && (
          <ul>
            {repos.map((rep) => <li key={rep.id}>{rep.name}</li>)}
          </ul>
        )}
      </div>
    </>
  );
};

export default ReposPage;
