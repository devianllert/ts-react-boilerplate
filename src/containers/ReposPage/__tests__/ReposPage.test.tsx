import React from 'react';
import { cleanup, fireEvent, act } from '@testing-library/react';

import { renderWithProviders } from '../../../utils/testUtils';

import ReposPage from '../ReposPage';
import * as reposService from '../../../services/repos.service';

import '@testing-library/jest-dom/extend-expect';

jest.mock('../../../services/repos.service');

describe('<Repos />', (): void => {
  afterEach(cleanup);

  it('should render and match the snapshot', (): void => {
    const { container } = renderWithProviders(<ReposPage />);

    expect(container).toMatchSnapshot();
  });

  it('should call fetch', (): void => {
    const { getByText } = renderWithProviders(<ReposPage />);
    const fetchSpy = jest.spyOn(reposService, 'fetchRepos');

    act(() => {
      fireEvent.click(getByText('Fetch'));
    });

    expect(fetchSpy).toHaveBeenCalled();
  });
});
