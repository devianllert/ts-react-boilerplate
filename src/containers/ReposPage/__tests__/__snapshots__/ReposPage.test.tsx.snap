// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repos /> should render and match the snapshot 1`] = `
<div>
  <div
    class="content"
  >
    <input
      name="repository"
      placeholder="Enter repository name"
      type="text"
      value=""
    />
    <button
      class="button primary medium"
      title="Fetch repositories"
      type="button"
    >
      Fetch
    </button>
  </div>
</div>
`;
